lib/collection/src/collection_manager/holders/segment_holder.rs

```diff
@@ -174,34 +177,47 @@ impl<'s> SegmentHolder {
 
     fn generate_new_key(&self) -> SegmentId {
         let key = thread_rng().gen::<SegmentId>();
-        if self.segments.contains_key(&key) {
+        if self.appendable_segments.contains_key(&key)
+            || self.non_appendable_segments.contains_key(&key)
+        {
             self.generate_new_key()
         } else {
             key
         }
     }
 
+    fn _insert(&mut self, segment: LockedSegment) -> SegmentId {
+        let key = self.generate_new_key();
+        if segment.get().read().is_appendable() {
+            self.appendable_segments.insert(key, segment);
+        } else {
+            self.non_appendable_segments.insert(key, segment);
+        }
+        key
+    }
```

Comment:
What about putting this just in `add()` rather than in a separate internal method. The function signature is the same, `add()` doesn't have any special logic.

Reply:
:thinking: